sparse/src/KokkosSparse_SortCrs.hpp

```diff
@@ -566,88 +584,171 @@ void sort_crs_graph(const crsGraph_t& G) {
+      numRows + 1);
   size_type numCompressedEntries = 0;
-  Kokkos::parallel_reduce(range_t(exec, 0, A.numRows()),
-                          Impl::MergedRowmapFunctor<rowmap_t, entries_t>(
-                              mergedRowmap, A.graph.row_map, A.graph.entries),
+  Kokkos::parallel_reduce(range_t(exec, 0, numRows),
+                          Impl::MergedRowmapFunctor<nc_rowmap_t, entries_t>(
+                              nc_rowmap_out, rowmap_in, entries_in),
                           numCompressedEntries);
+  if (nnz == numCompressedEntries) {
+    // No merges to do, so just return A. Save the time of allocating and
+    // filling a copy.
+    if constexpr (std::is_const_v<typename rowmap_t::value_type>) {
+      rowmap_out = rowmap_in;
+    } else {
+      // rowmap_t is non-const, so we can't directly assign rowmap_in to
+      // rowmap_out. Forced to deep copy it to maintain const-correctness.
+      Kokkos::deep_copy(exec, nc_rowmap_out, rowmap_in);
+      rowmap_out = nc_rowmap_out;
+    }
+    entries_out = entries_in;
+    values_out  = values_in;
+    return;
+  }
   // Prefix sum to get rowmap
-  KokkosKernels::Impl::kk_exclusive_parallel_prefix_sum<rowmap_t, exec_space>(
-      exec, A.numRows() + 1, mergedRowmap);
-  entries_t mergedEntries(Kokkos::view_alloc(exec, "SortedMerged entries"),
+  KokkosKernels::Impl::kk_exclusive_parallel_prefix_sum<nc_rowmap_t,
+                                                        exec_space>(
+      exec, numRows + 1, nc_rowmap_out);
+  rowmap_out  = nc_rowmap_out;
+  entries_out = entries_t(Kokkos::view_alloc(exec, Kokkos::WithoutInitializing,
+                                             "SortedMerged entries"),
                           numCompressedEntries);
-  values_t mergedValues(Kokkos::view_alloc(exec, "SortedMerged values"),
+  values_out  = values_t(Kokkos::view_alloc(exec, Kokkos::WithoutInitializing,
+                                           "SortedMerged values"),
                         numCompressedEntries);
   // Compute merged entries and values
   Kokkos::parallel_for(
-      range_t(exec, 0, A.numRows()),
-      Impl::MatrixMergedEntriesFunctor<c_rowmap_t, entries_t, values_t>(
-          A.graph.row_map, A.graph.entries, A.values, mergedRowmap,
-          mergedEntries, mergedValues));
-  // Finally, construct the new compressed matrix
-  return crsMat_t("SortedMerged", A.numRows(), A.numCols(),
-                  numCompressedEntries, mergedValues, mergedRowmap,
-                  mergedEntries);
+      range_t(exec, 0, numRows),
+      Impl::MatrixMergedEntriesFunctor<rowmap_t, entries_t, values_t>(
+          rowmap_in, entries_in, values_in, rowmap_out, entries_out,
+          values_out));
 }
 
-template <typename crsMat_t>
-crsMat_t sort_and_merge_matrix(const crsMat_t& A) {
-  return sort_and_merge_matrix(typename crsMat_t::execution_space(), A);
+template <typename exec_space, typename rowmap_t, typename entries_t,
```

Comment:
If the user does not pass an execution space, should we simply use that of A instead of asking for an extra template parameter?
I think that would be more in line with the interfaces that use a CrsMat or CrsGraph as input.

Reply:
@lucbv Done, now the exec space template parameter is optional for all the functions. These are all equivalent now:
```
sort_crs_graph(rowmap, entries);
sort_crs_graph<exec_space>(rowmap, entries);
sort_crs_graph<exec_space, rowmap_t, entries_t>(rowmap, entries);
```